tests/_runner.lua

```diff
@@ -43,7 +43,54 @@ runner.step_kill_clients = function(step)
     end
 end
 
-runner.run_steps = function(steps, options)
+--- Print a message if verbose mode is enabled
+-- @tparam string message The message to print
+function runner.verbose(message)
+    if verbose then
+        io.stderr:write(message .. "\n")
+    end
+end
+
+--- When using run_direct(), this function indicates that the test is now done.
+-- @tparam[opt=nil] string message An error message explaining test failure.
+function runner.done(message)
+    if message then
+        io.stderr:write("Error: " .. message .. "\n")
+        if not runner.quit_awesome_on_error then
+            io.stderr:write("Keeping awesome open...\n")
+            return
+        end
+    end
+
+    local client_count = #client.get()
+    if client_count > 0 then
+        io.stderr:write(string.format(
+            "NOTE: there were %d clients left after the test.\n", client_count))
+
+        -- Remove any clients.
+        for _,c in ipairs(client.get()) do
+            c:kill()
+        end
+    end
+
+    if not message then
+        io.stderr:write("Test finished successfully.\n")
+    end
+    awesome.quit()
+end
+
+--- This function is called to indicate that a test does not use the run_steps()
+-- facility, but instead runs something else directly.
+function runner.run_direct()
```

Comment:
s/run_direct/is_running/ maybe?
The current name indicated that it would do something, although with the example in test-gravity it is called only after the test - and it seems like the `runner.run_direct()` call is not necessary there after all?!
It could also just set `runner.is_running` instead maybe.

Reply:
Well, no. `test-gravity.lua` starts an external process and afterwards it calls `run_direct()`. However, at this time, the test is not yet done. Only when the external process finishes some time later is `done()` actually called. So, it's still asynchronous.

The idea why I call `run_direct()` at the end is because it "disarms" a timer in `_runner.lua` that would otherwise complain that the test was not started correctly and quit awesome. So, if starting the external process would fail, awesome would quit relatively quickly instead of waiting for the 180 seconds timeout.

Right now I don't see a nice way to make this more explicit... well, `run_direct` could get a function as its argument which it just runs, but that might also be weird...

Ideas?

Reply:
Thanks for the detailed explanation, as always.

It is fine by me.